en/202407/exploring-geo-distribution-in-emqx-for-enhanced-scalability.md

```diff
@@ -0,0 +1,258 @@
+
+This feature is not yet enabled by default, but it's pretty easy to turn on.
+
+```
+broker.routing.batch_sync.enable_on = replicant
+```
+
+This snippet enables batch synchronization only on replicants. Setting it to `all` would enable it on all nodes, and in general will likely be beneficial for a wide variety of workloads, mostly thanks to the broker pool being freed up from doing extra synchronization work. In our case, the effects would not be as pronounced and harder to illustrate.
+
+The default is `none`, and the reason for that is safety. We want to provide smooth rolling upgrade experience, and enabling it *before* rolling out EMQX 5.7.0 over, let's say, 5.4.1 could lead to some temporary unavailability. If you find yourself upgrading from EMQX 5.5.0 or later however, you should not worry about enabling it right away.
+
+Ok then, let's see how it goes now.
+
+```shell
+$ mqttx bench sub -h ape1.emqx.dev --count 1000 --interval 0 --topic t/mqttx/+/%i/sub/+/#
+[11:46:21 PM] › ℹ  Start the subscribe benchmarking, connections: 1000, req interval: 0ms, topic: t/mqttx/+/%i/sub/+/#
+✔  success   [1000/1000] - Subscribed to t/mqttx/+/1000/sub/+/#
+[11:46:24 PM] › ℹ  Created 1000 connections in 2.585s
+```
+
+Pretty significant improvement. The network throughput is now used more efficiently, which in turn translates to better `SUBSCRIBE` throughput. Roughly the same effect could most likely be achieved by inflating process pool at the cost of more bookkeeping and increased memory usage, but this new feature is more flexible and easier to use.
+
+Why are we still far from the theoretical limit of ~600 milliseconds? Most plausible explanation is how occasional network hiccups are handled. A link may be congested for a quick moment, that would cause a packet or few to be lost, and the TCP stack would have to retransmit them once *RTO (retransmit timeout)* passes. This is usually barely noticeable in an already established connection, but there's often a huge difference when an initial SYN packet is lost. In this case, the TCP stack had no chance to estimate the optimal RTO yet, so it would have to start with a conservative default of *1 second*. Coming back to our workload, it's likely that while most of the clients were able to connect and subscribe in a fraction of a second, a couple of them tripped over SYN packet loss, which in turn caused the total time to increase by a second. Or, if we were extremely unlucky, two seconds.
```

Comment:
```suggestion
Why are we still far from the theoretical limit of ~600 milliseconds? The most plausible explanation is how occasional network hiccups are handled. A link may be congested for a quick moment, which would cause a packet or few to be lost, and the TCP stack would have to retransmit them once *RTO (retransmit timeout)* passes. This is usually barely noticeable in an already established connection, but there's often a huge difference when an initial SYN packet is lost. In this case, the TCP stack had no chance to estimate the optimal RTO yet, so it would have to start with a conservative default of *1 second*. Coming back to our workload, it's likely that while most of the clients were able to connect and subscribe in a fraction of a second, a couple of them tripped over SYN packet loss, which in turn caused the total time to increase by a second. Or, if we were extremely unlucky, two seconds.
```